### ISSUE TITLE
GitHub action to do Deployment for you

### ISSUE THREAD

**Issue by pryrt** [notepad-plus-plus/npp-usermanual#547]
@donho,

What do you think about having a GitHub action that would perform the "deployment" (publishing to the website) for you?  We could set it up so that it automatically publishes any time a Tag is made.  This would mean that I wouldn't have to keep pestering you every time the manual is ready for the next release, freeing up more of your time for focusing on other things. :-)

**Details**
With using SSH keys stored in encrypted GitHub variables, a GitHub action would be able to copy the bundle to the npp-user-manual.org host machine, and run the command to extract the new version into the correct directory on the server.

If you were okay with me pursuing this, the steps would be:
1. You let me know what command(s) you use to unbundle, and what directory it goes into
2. I would do experimental deployments to a machine I have access to, and set up the GitHub Action to work correctly
3. I would then give you the instructions I'd figured out for correctly creating and storing the SSH key, so that you could do that part
4. From then on, I should be able to trigger a Tag here, and have it automatically publish the User Manual for us.

If you aren't comfortable with automating that, I understand, and I'll just keep pestering you every once in a while to manually do the next User Manual release. :-)

**Comment by donho** [comment]
@pryrt 
It sounds interesting. But some points are not clear to me.

The procedure for the moment is:
1. You assign me a job via GitHub, the I sync my local master from the one of  GitHub.
2. I increase the version, and tag a release on GitHub.
3. I generate whole site into HTML from Markdown locally.
4. I login (with password) into Npp User Manual host server via FTP, and copy the generated content.

Could you provide the procedure/schema of the automatic deployment please?

**Comment by pryrt** [comment]
@donho,

Based on that description of the current procedure, the modified procedure would be:
1. One of the User Manual Admin [team](https://github.com/orgs/notepad-plus-plus/teams/notepad-user-manual-admin) (usually me) will increase the version and tag a release on GitHub
2. GitHub action ("GHA") will run because of the tag
    1. GHA will generate the HTML from the Markdown
    2. GHA would login to the Npp User Manual host server using SFTP and an SSH key (which is stored in a protected GitHub variable, so that no outsiders can read or use the key), and copy the generated content

**Comment by donho** [comment]
Using GH's private & public pair key might be good.
However, for login SSH, GH needs rather public key generated from the host of Npp User Manual site, right?

**Comment by pryrt** [comment]
I believe it's actually the other way around, because it's GH trying to access the usermanual server, so GH gets the private key and the usermanual server gets the public key.  But if you want to go down this route, I will experiment with a remote server that I have credentials for, and will figure out where each key goes for the setup to work.  When I know everything is working, I can give the right steps for which part of the key you need to put where.  (For making sure I'm replicating things correctly, it would be useful to know what directory the HTML files go into on the usermanual server.)

**Comment by donho** [comment]
> because it's GH trying to access the usermanual server, so GH gets the private key and the usermanual server gets the public key.

Oups, you're right.
OK, we can try at least.
Please let me know what you need.

**Comment by pryrt** [comment]
Okay, I've got a setup that I've tested as working with my dummy site.

1. Create an openssh private/public key pair (with no password on the key).  A command like the following will work from a linux prompt or from git bash:
    ```
    ssh-keygen -t ed25519 -C "UserManual Deployment Key 2023-Aug-17" -f ./id_ed25519_usermanual
    ```
2. copy the contents of the public key (.pub file it just created), and go to the npp-user-manual.org host, edit the file `~/.ssh/authorized_keys` file 
    - if the file doesn't exist, create it with 0644 permissions
    - if the directory doesn't exist, create it with 0700 permissions
3. copy the contents of the private key into the clipboard
4. go to https://github.com/notepad-plus-plus/npp-usermanual/settings/secrets/actions
   - **New Repository Secret** => name: `USERMANUAL_KEY` , and paste the private key as the contents, **Add Secret**
   - **New Repository Secret** => name: `USERMANUAL_USERNAME`, enter the username you use to log into the server as the value, **Add Secret**
   - **New Repository Secret** => name: `USERMANUAL_HOST`, enter the hostname (presumably `npp-user-manual.org`, but it's whatever you normally put in your FTP client as the hostname), **Add Secret**

Once those are added, reply here, and let me know what directory on the host holds the website, so that I can set up the FTP command to put the HTML in the right directory.  (I am guessing it's probably something like `.../www/` or `.../public_html/` off of your home directory, but I need to know for sure in order to correctly.)

Once you've got that ready, I will bring in the new GitHub action, and verify that I can commit and publish.  I'll let you know if I need anything after that, or I will close this issue if it works right.

Thank you.

**Comment by donho** [comment]
@pryrt 
Sorry for the late reply.

> copy the contents of the public key (.pub file it just created), and go to the npp-user-manual.org host, edit the file ~/.ssh/authorized_keys file 

By `~/`, which user's home do you mean? Currently the ower of `www` directory is root, but the root account has been disabled at the beginning for the reason of security (and I use another account XXX which has all right). What's your suggestion?

**Comment by pryrt** [comment]
It would be the home directory for the XXX account (whatever account it is you use to log in via FTP or other means)

**Comment by donho** [comment]
> name: USERMANUAL_HOST, enter the hostname (presumably npp-user-manual.org, but it's whatever you normally put in your FTP client as the hostname), Add Secret

It's SFTP, so I will provide like `23.147.256.11:1313` as content of USERMANUAL_HOST. Is it OK for you?

**Comment by pryrt** [comment]
Port is actually a separate parameter to the GitHub action that I'll be using.  If it's not using the default SFTP port 22, please put the port number separately in USERMANUAL_PORT (so your example would be USERMANUAL_HOST=`23.147.256.11` and USERMANUAL_PORT=`1313`)

**Comment by donho** [comment]
@pryrt 
> Once those are added, reply here, and let me know what directory on the host holds the website, so that I can set up the FTP command to put the HTML in the right directory.

It's done from my side. You go now.
The directory is classic:
`/var/www/npp-user-manual.org/public_html/`

Please let me posted after your tests.

**Comment by pryrt** [comment]
The tests have worked.  I was able to use the GitHub Action workflow to regenerate the HTML and write it to the `/var/www/npp-user-manual.org/public_html/` .  Looking at view-source:https://npp-user-manual.org/docs/history/ , I can see the HTML comment I added on the live website.

Once #550 is merged, any time one of the User Manual Admins group creates a new tag, the Action will automatically write it to the website.

Thanks for setting up the credentials for this to work.

**Comment by donho** [comment]
Beautiful! 
So #550 is all we need for the automatic deployment? By creating a new tag, you mean creating a release in GitHub?

**Comment by pryrt** [comment]
> So #550 is all we need for the automatic deployment? 

Correct

> By creating a new tag, you mean creating a release in GitHub?

Yes

**Comment by pryrt** [comment]
2024 Notes: As Don and I tried to get the deployment running on the updated server this week, we ran across a couple of things to note, both mentioned in https://github.com/wlixcc/SFTP-Deploy-Action/issues/1#issuecomment-777877004
1. If you are using ed25519, it _requires_ a newline at the end.  My experiments confirmed this worked with my personal website host, but Don wasn't able to get the ed25519 to work on the user manual host.
2. newer `ssh-keygen` no longer default to PEM format for the private key, which is required for the SFTP-Deploy-Action 

After learning these lessons, Don was able to get it to work this time using the RSA keytype and PEM output format, using the command:
```
ssh-keygen -t rsa -m PEM -C "UserManual Deployment Key 2024-07-17" -f ./id_rsa_PEM_usermanual
```

Also, the new host setup requires a relative path for the destination, instead of an absolute path.  Trying the `~/.../` notation doesn't work.  In the end, he got it working by changing the SFTP-Deploy-Action config to use `remote_path: 'nppusermanual/public_html/'` to make it relative to the username's default home directory

Hopefully, these notes will help it go smoothly if we need to change the setup again in the future.
